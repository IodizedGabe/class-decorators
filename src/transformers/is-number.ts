import { composePropertyDecorators, IsNumberOptions } from '@hippo-oss/dto-decorators';
import { Type } from 'class-transformer';

import { createBasePropertyDecorators } from './base';

export function IsNumber(options: IsNumberOptions = {}): PropertyDecorator {
    return composePropertyDecorators([
        ...createBasePropertyDecorators(options),

        // convert string to number
        Type(() => Number),
    ]);
}
