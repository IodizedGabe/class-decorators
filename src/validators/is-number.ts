import { composePropertyDecorators, IsNumberOptions } from '@hippo-oss/dto-decorators';
import { IsNumber as IsNumberDecorator, Max, Min } from 'class-validator';

import { createBasePropertyDecorators } from './base';

export function IsNumber(options: IsNumberOptions): PropertyDecorator {
    return composePropertyDecorators([
        ...createBasePropertyDecorators(options),

        // validate data as a number
        IsNumberDecorator({}, { each: !!options.isArray }),

        // maybe: add a maximum value
        options?.maxValue !== undefined ? Max(options.maxValue, { each: !!options.isArray }) : undefined,

        // maybe: add a minimum value
        options?.minValue !== undefined ? Min(options.minValue, { each: !!options.isArray }) : undefined,
    ]);
}
