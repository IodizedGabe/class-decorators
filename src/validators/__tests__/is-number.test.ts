import { validateSync } from 'class-validator';

import { IsNumber } from '../is-number';

describe('IsNumber', () => {
    it('allows optional number to be omitted', () => {
        class Fixture {
            @IsNumber({
                optional: true,
            })
            property!: number;
        }

        const obj = new Fixture();
        const errors = validateSync(obj);
        expect(errors).toHaveLength(0);
    });
    it('allows nullable number to be omitted', () => {
        class Fixture {
            @IsNumber({
                nullable: true,
            })
            property!: number;
        }

        const obj = new Fixture();
        const errors = validateSync(obj);
        expect(errors).toHaveLength(0);
    });
    it('does not allow required number to be omitted', () => {
        class Fixture {
            @IsNumber({})
            property!: number;
        }

        const obj = new Fixture();
        const errors = validateSync(obj);
        expect(errors).toHaveLength(1);
        expect(errors).toMatchSnapshot();
    });
    it('does not allow non-number', () => {
        class Fixture {
            @IsNumber({})
            property!: number;
        }

        const obj = new Fixture();
        obj.property = '42' as unknown as number;

        const errors = validateSync(obj);
        expect(errors).toHaveLength(1);
        expect(errors).toMatchSnapshot();
    });
    it('does not allow number that is too small', () => {
        class Fixture {
            @IsNumber({
                minValue: 0,
            })
            property!: number;
        }

        const obj = new Fixture();
        obj.property = -1.25;

        const errors = validateSync(obj);
        expect(errors).toHaveLength(1);
        expect(errors).toMatchSnapshot();
    });
    it('does not allow number that is too large', () => {
        class Fixture {
            @IsNumber({
                maxValue: 0,
            })
            property!: number;
        }

        const obj = new Fixture();
        obj.property = 1.25;

        const errors = validateSync(obj);
        expect(errors).toHaveLength(1);
        expect(errors).toMatchSnapshot();
    });
    it('allows number', () => {
        class Fixture {
            @IsNumber({
                maxValue: 2,
                minValue: 0,
            })
            property!: number;
        }

        const obj = new Fixture();
        obj.property = 1.25;

        const errors = validateSync(obj);
        expect(errors).toHaveLength(0);
    });
});
