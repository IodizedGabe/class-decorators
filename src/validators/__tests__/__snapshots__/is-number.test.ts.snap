// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IsNumber does not allow non-number 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isNumber": "property must be a number conforming to the specified constraints",
    },
    "property": "property",
    "target": Fixture {
      "property": "42",
    },
    "value": "42",
  },
]
`;

exports[`IsNumber does not allow number that is too large 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "max": "property must not be greater than 0",
    },
    "property": "property",
    "target": Fixture {
      "property": 1.25,
    },
    "value": 1.25,
  },
]
`;

exports[`IsNumber does not allow number that is too small 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "min": "property must not be less than 0",
    },
    "property": "property",
    "target": Fixture {
      "property": -1.25,
    },
    "value": -1.25,
  },
]
`;

exports[`IsNumber does not allow required number to be omitted 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "property should not be null or undefined",
      "isNumber": "property must be a number conforming to the specified constraints",
    },
    "property": "property",
    "target": Fixture {},
    "value": undefined,
  },
]
`;
